File: java/validsudoku.java
public class Solution {
    
    char[][] board;
    
    public boolean isValidSudoku(char[][] board) {
        // Start typing your Java solution below
        // DO NOT write main() function
        this.board = board;
        for(int i=0;i<9;i++){
            for(int j=0;j<9;j++){
                if(board[i][j]!='.'){
                    int number = board[i][j]-'0';
                    board[i][j] = '.';
                    if(!check(i,j,number)){
                        return false;
                    }
                    board[i][j] = (char)(number+'0');
                }
            }
        }
        return true;
    }

    private boolean check(int x, int y, int number){
        char target = (char) (number+'0');
        for(int i=0;i<9;i++){
            if(board[x][i]==target){
                return false;
            }
            if(board[i][y]==target){
                return false;
            }
        }
        for(int i=(x/3)*3;i<=(x/3)*3+2;i++){
            for(int j=(y/3)*3;j<=(y/3)*3+2;j++){
                if(board[i][j]==target){
                    return false;
                }
            }
        }
        return true;
    }

    
}

File: cpp/ValidSudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char> > &board) {
        bool grid[3][3][9];
        bool row[9][9];
        bool col[9][9];
        memset(grid,0,sizeof(grid));
        memset(row,0,sizeof(row));
        memset(col,0,sizeof(col));
        for(int i=0;i<9;i++) {
            for(int j=0;j<9;j++) {
                if(board[i][j]=='.') {
                    continue;
                }
                int number = board[i][j]-'0';
                if(grid[i/3][j/3][number]) {
                    return false;
                }
                grid[i/3][j/3][number] = true;
                if(row[i][number]){
                    return false;
                }
                row[i][number] = true;
                if(col[j][number]){
                    return false;
                }
                col[j][number] = true;
            }
        }
        return true;
    }
};
